packages/grafana-ui/src/types/datasource.ts

```diff
@@ -188,28 +188,9 @@ export abstract class DataSourceApi<
   init?: () => void;
 
   /**
-   * Query for data, and optionally stream results to an observer.
-   *
-   * Are you reading these docs aiming to execute a query?
-   * +-> If Yes, then consider using panelQueryRunner/State instead.  see:
-   *  * {@link https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryRunner.ts PanelQueryRunner.ts}
-   *  * {@link https://github.com/grafana/grafana/blob/master/public/app/features/dashboard/state/PanelQueryState.ts PanelQueryState.ts}
-   *
-   * If you are implementing a simple request-response query,
-   * then you can ignore the `observer` entirely.
-   *
-   * When streaming behavior is required, the Promise can return at any time
-   * with empty or partial data in the response and optionally a state.
-   * NOTE: The data in this initial response will not be replaced with any
-   * data from subsequent events. {@see DataStreamState}
-   *
-   * The request object will be passed in each observer callback
-   * so the callback could assert that the correct events are streaming and
-   * unsubscribe if unexpected results are returned.
+   * Query for data, and optionally stream results
    */
-  abstract query(request: DataQueryRequest<TQuery>, observer?: DataStreamObserver): Promise<DataQueryResponse>;
-
-  observe?(request: DataQueryRequest<TQuery>): Observable<DataQueryResponsePacket>;
+  abstract query(request: DataQueryRequest<TQuery>): Promise<DataQueryResponse> | Observable<DataQueryResponse>;
```

Comment:
I like that we merged things into a single function, but don't like that the response is `Observable<DataQueryResponse>`  that is pretty close to `Observable<{data:any[]>`

Can we do:
```
Promise<DataQueryResponse> | Observable<DataQueryPacket>
```
and at 7.0 deprecate `Promise<DataQueryResponse>`.  Then we can have more clear typed response.

Reply:
One alternative would be to modify DataQueryResponse like this:

```
 export interface DataQueryResponse {
  data: LegacyResponseData[];
  frames: (DataFrame | DataFrameDTO)[]
  key?: string;
}
```

Reply:
Thinking more... I think we should keep:
```
export interface DataQueryResponse {
  data?: (LegacyResponseData | DataFrame | DataFrameDTO)[];
  key?: string;
}
```

At 7.0, we could change that to:
```
export interface DataQueryResponse {
  data?: (DataFrame | DataFrameDTO)[];
  key?: string;
}
```
BUT still support LegacyResponseData.  By "support" I mean we should still wrap all results in a function that calls `toDataFrame(any)`

Reply:
sounds like a plan